google-cloud-firestore/src/main/java/com/google/cloud/firestore/UserDataConverter.java

```diff
@@ -227,4 +228,45 @@ static Object decodeValue(FirestoreRpcContext<?> rpcContext, Value v) {
             String.format("Unknown Value Type: %s", typeCase));
     }
   }
+
+  static Object decodeGoogleProtobufValue(com.google.protobuf.Value v) {
+    switch (v.getKindCase()) {
+      case NULL_VALUE:
+        return null;
+      case BOOL_VALUE:
+        return v.getBoolValue();
+      case NUMBER_VALUE:
+        return v.getNumberValue();
+      case STRING_VALUE:
+        return v.getStringValue();
+      case LIST_VALUE:
+        List<Object> list = new ArrayList<>();
+        List<com.google.protobuf.Value> lv = v.getListValue().getValuesList();
+        for (com.google.protobuf.Value iv : lv) {
+          list.add(decodeGoogleProtobufValue(iv));
+        }
+        return list;
+      case STRUCT_VALUE:
+        Map<String, Object> outputMap = new HashMap<>();
+        Map<String, com.google.protobuf.Value> inputMap = v.getStructValue().getFieldsMap();
+        for (Map.Entry<String, com.google.protobuf.Value> entry : inputMap.entrySet()) {
+          outputMap.put(entry.getKey(), decodeGoogleProtobufValue(entry.getValue()));
+        }
+        return outputMap;
+      default:
+        throw FirestoreException.forInvalidArgument(
+            String.format("Unknown Value Type: %s", v.getKindCase().getNumber()));
+    }
+  }
+
+  static Map<String, Object> decodeStruct(@Nullable Struct struct) {
+    Map<String, Object> result = new HashMap<>();
+    if (struct == null || struct.getFieldsCount() == 0) {
+      return result;
+    }
+    for (Map.Entry<String, com.google.protobuf.Value> entry : struct.getFieldsMap().entrySet()) {
+      result.put(entry.getKey(), decodeGoogleProtobufValue(entry.getValue()));
+    }
+    return result;
```

Comment:
Likewise here, I think Gauva map view is better.

```suggestion
    if (struct == null || struct.getFieldsCount() == 0) {
      return Collections.emptyMap();
    }
    return Maps.transformValues(
        struct.getFieldsMap(),
        UserDataConverter::decodeGoogleProtobufValue
    );
```

Reply:
Done.